File: decodestring.go
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
have 2 stacks one for number, one for string
and curr string to return
loop each element of string
if its a number then loop again till we get all numbers and finally append it to the number stack
else if its [ then append the curr to stack and reset curr to ""
else if its ] then pop number and string from stack and repeat curr popped number times and append popped string to it finally append all to curr
else if its a alphabet then cppend to curr
*/
package main

import "fmt"

func decodeString(s string) string {
	stNum := []int{}
	stStr := []string{}
	curr := ""
	x := 0
	for x<len(s) {
		if s[x] >= '0' && s[x] <= '9' {
			num :=0
			for x<len(s) && s[x] >= '0' && s[x] <= '9' {
				num = num * 10 + int(s[x] - '0')
				x++
			}
			stNum = append(stNum, num)
		} else if s[x] == '[' {
			stStr = append(stStr, curr)
			curr = ""
			x++
		} else if s[x] == ']' {
			currString := stStr[len(stStr) - 1]
			stStr = stStr[:len(stStr) - 1]

			currCount := stNum[len(stNum) - 1]
			stNum = stNum[:len(stNum) - 1]

			finStr := ""
			for y:=0;y<currCount;y++{
				finStr += curr
			}
			curr = currString + finStr
			x++
		} else {
			curr+=string(s[x])
			x++
		}
	}
	return curr
}

func MainDecodeString() {
	fmt.Println(decodeString("3[a]2[bc]")) //expected aaabcbc
}

File: decode-string.js
// Time Complexity : O(N)
// Space Complexity : O(B) where B is the number of opening brackets
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : The recursive solution requires a global variable, which messes with the
// JavaScript compiler on LeetCode.  I got it to work by resetting the variable before the last return statement (line 70).

/**
 * @param {string} s
 * @return {string}
 */

// Iterative
// var decodeString = function(s) {
//     if (s == null || s.length == 0) return '';
    
//     let stack = [],
//         currNum = 0,
//         currStr = '';
//     for (let i = 0; i < s.length; i++) {
//         let c = s[i];
//         if (parseInt(c) == c) {
//             currNum = currNum * 10 + parseInt(c);
//         } else if (c == '[') {
//             stack.push([currStr, currNum])
//             currNum = 0;
//             currStr = '';
//         } else if (c == ']') {
//             let stackPop = stack.pop(),
//                 times = stackPop[1],
//                 oldString = stackPop[0],
//                 newString = '';
//             for (let j = 0; j < times; j++) {
//                 newString += currStr;
//             }
//             currStr = oldString + newString;
//         } else {
//             currStr += c;
//         }
//     }
//     return currStr;
// };

// Recursive
let i = 0;
var decodeString = function(s) {
    if (s == null || s.length == 0) return '';
    
    let result = '',
        num = 0;
    while(i < s.length) {
        let c = s[i];
        if (parseInt(c) == c) {
            i++;
            num = num * 10 + parseInt(c);
        } else if (c == '[') {
            i++;
            let inner = decodeString(s);
            for (let j = 0; j < num; j++) {
                result += inner;
            }
            num = 0;
        } else if (c == ']') {
            i++;
            return result;
        } else {
            i++;
            result += c;
        }
    }
    i = 0;
    return result;
};
